### PR TITLE
[SPARK-22745][SQL] read partition stats from Hive

### PR DESCRIPTION
**Pull request by wzhfy**
## What changes were proposed in this pull request?

Currently Spark can read table stats (e.g. `totalSize, numRows`) from Hive, we can also support to read partition stats from Hive using the same logic.

## How was this patch tested?

Added a new test case and modified an existing test case.

### CODE REVIEW

**Comment by wzhfy** [comment]
cc @cloud-fan @gatorsmile

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 1]
```diff
@@ -413,32 +413,7 @@ private[hive] class HiveClientImpl(
         case (key, _) => excludedTableProperties.contains(key)
       }
       val comment = properties.get("comment")
-
-      // Here we are reading statistics from Hive.
-      // Note that this statistics could be overridden by Spark's statistics if that's available.
-      val totalSize = properties.get(StatsSetupConst.TOTAL_SIZE).map(BigInt(_))
```
Comment (wzhfy) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: The code path is moved to the method `readHiveStats`

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala [hunk 2]
```diff
@@ -213,6 +213,29 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
     }
   }
 
+  test("SPARK- - read Hive's statistics for partition") {
```
Comment (wangyum) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: SPARK- -> SPARK-22745?

Comment (wzhfy) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: oh, I forgot it, thanks!

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala [hunk 3]
```diff
@@ -353,15 +374,6 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
       createPartition("2010-01-02", 11,
         "SELECT '1', 'A' from src UNION ALL SELECT '1', 'A' from src")
 
-      sql(s"ANALYZE TABLE $tableName PARTITION (ds='2010-01-01') COMPUTE STATISTICS NOSCAN")
-
-      assertPartitionStats("2010-01-01", "10", rowCount = None, sizeInBytes = 2000)
-      assertPartitionStats("2010-01-01", "11", rowCount = None, sizeInBytes = 2000)
-      assert(queryStats("2010-01-02", "10") === None)
-      assert(queryStats("2010-01-02", "11") === None)
```
Comment (wzhfy) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: After the change, these checks are not right as we read hive stats. So I remove them.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 4]
```diff
@@ -1021,8 +998,38 @@ private[hive] object HiveClientImpl {
...
+
+  /**
+   * Reads statistics from Hive.
+   * Note that this statistics could be overridden by Spark's statistics if that's available.
+   */
+  private def readHiveStats(properties: Map[String, String]): Option[CatalogStatistics] = {
+    val totalSize = properties.get(StatsSetupConst.TOTAL_SIZE).map(BigInt(_))
+    val rawDataSize = properties.get(StatsSetupConst.RAW_DATA_SIZE).map(BigInt(_))
+    val rowCount = properties.get(StatsSetupConst.ROW_COUNT).map(BigInt(_))
+    // TODO: check if this estimate is valid for tables after partition pruning.
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: do we still need this TODO?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 5]
```diff
@@ -413,32 +413,7 @@ private[hive] class HiveClientImpl(
...
-        // zero after INSERT command. So they are used here only if they are larger than zero.
-        if (totalSize.isDefined && totalSize.get > 0L) {
-          Some(CatalogStatistics(sizeInBytes = totalSize.get, rowCount = rowCount.filter(_ > 0)))
-        } else if (rawDataSize.isDefined && rawDataSize.get > 0) {
-          Some(CatalogStatistics(sizeInBytes = rawDataSize.get, rowCount = rowCount.filter(_ > 0)))
-        } else {
-          // TODO: still fill the rowCount even if sizeInBytes is empty. Might break anything?
-          None
-        }
+      val hiveStats = readHiveStats(properties)
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: nit: we can inline it

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 6]
```diff
@@ -1011,6 +986,8 @@ private[hive] object HiveClientImpl {
    */
   def fromHivePartition(hp: HivePartition): CatalogTablePartition = {
     val apiPartition = hp.getTPartition
+    val properties: Map[String, String] =
+      if (hp.getParameters != null) hp.getParameters.asScala.toMap else Map.empty
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: nit: if can't fit in one line, prefer
```
val xxx = if {
  ...
} else {
  ...
}
```

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala [hunk 7]
```diff
@@ -213,6 +213,27 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
...
+
+      assert(partition.stats.get.sizeInBytes == 5812)
+      assert(partition.stats.get.rowCount.isEmpty)
+
+      hiveClient
+        .runSqlHive(s"ANALYZE TABLE $tableName PARTITION (ds='2017-01-01') COMPUTE STATISTICS")
+      partition = spark.sessionState.catalog
+        .getPartition(TableIdentifier(tableName), Map("ds" -> "2017-01-01"))
+
+      assert(partition.stats.get.sizeInBytes == 5812)
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: I'm expecting `totalSize` is picked here and the `sizeInBytes` would be changed, did I miss something?

Comment (wzhfy) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: `totalSize` exists after the INSERT INTO command, so here `sizeInBytes` doesn't change after ANALYZE command, only rowCount is added.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala (hunk 4 above)
Comment (wzhfy) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: good catch, we can remove this

**Comment by cloud-fan** [comment]
LGTM

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**